### PR TITLE
Refactor media modules

### PR DESCRIPTION
**Pull request by clokep**
This makes a few changes to the media files that I've wanted to do forever:

* Removes the `v1` directory from  `test.rest.media.v1`.
* Moves the non-REST stuff *out* of `synapse.rest.media.v1` (to `synapse.media`) (A similar change is made to tests so they end up in `tests.media`.)
* Remove the `v1` directory from `synapse.rest.media`, we can't do this completely though since media repositories (and spam checkers) refer to a few things here, see:

https://github.com/matrix-org/synapse-s3-storage-provider/blob/fa27fa1a92bcbeb42b10399641348bee0ddf2c72/s3_storage_provider.py#L31-L32

https://github.com/Telegraphenbauanstalt/synapse-mediacheck/blob/ca0506f8e533983e252222de1e19cbe600213376/synapse_mediacheck/synapse_mediacheck.py#L23-L24

### CODE REVIEW

**Comment by clokep** [comment]
See #10600 for prior art.

**Comment by clokep** [comment]
(I have plans to do more refactoring to this stuff, but wanted to start by just moving all the files.)

File: synapse/rest/media/v1/storage_provider.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# This exists purely for backwards compatibility with media providers.
+from synapse.media.storage_provider import StorageProvider  # noqa: F401
```
Comment (DMRobertson) [inline synapse/rest/media/v1/storage_provider.py:17]: Is this necessary in the light of https://github.com/matrix-org/synapse/pull/15146/files#diff-93cc7d2bb16bade8a31d19c3f919055b5c93490fb5ab935c5486aeb87178f335L181-R187 ? I suppose it is if some modules have been importing from here.

**Review by DMRobertson — APPROVED** [review]
LGTM!

File: synapse/rest/media/v1/storage_provider.py (hunk 1 above)
Comment (clokep) [reply synapse/rest/media/v1/storage_provider.py:17]: Yeah, pretty much. I'm not really sure what the "real" fix is there (and I think the S3 provider is the only 3rd party one, but I'm not really sure).

**Pull request merged**